File: luau/QuaternionSpring.lua
--!strict
--[[
    Source: https://github.com/probablytukars/LuaQuaternion
    Based on: https://github.com/Quenty/NevermoreEngine/tree/main/src/spring
    [MIT LICENSE]
]]

local Quaternion = require(script.Parent.Quaternion)
type Quaternion = Quaternion.Quaternion
local ERROR_FORMAT = "%q is not a valid member of QuaternionSpring."

local QuaternionSpring = {_type = "QuaternionSpring"}

type t_QuaternionSpring = {
	new: (initial: Quaternion, damping: number, speed: number, clock: () -> number) -> QuaternionSpring,
	Reset: (target: Quaternion?) -> nil,
	Impulse: (self: QuaternionSpring, velocity: Vector3) -> nil,
	TimeSkip: (self: QuaternionSpring, delta: number) -> nil,
	
	Position: Quaternion,
	p: Quaternion,
	Velocity: Vector3,
	v: Vector3,
	Target: Quaternion,
	t: Quaternion,
	Damping: number,
	d: number,
	Speed: number,
	s: number,
	Clock: () -> number,
	
	_clock: () -> number,
	_time: number,
	_position:  Quaternion,
	_velocity:  Vector3,
	_target:  Quaternion,
	_damping:  number,
	_speed:  number,
	_initial:  Quaternion,
}

export type QuaternionSpring = typeof(setmetatable({} :: t_QuaternionSpring, QuaternionSpring))

--[=[
    @class QuaternionSpring
    @grouporder ["Constructors", "Methods"]
    
    This class represents a rotational spring using Quaternions.
    Velocity is a Vector3 where the magnitude is the angle, and the unit
    is the axis (if angle is > 0).
    
    This is lazily evaluated meaning it only updates when indexed.
--]=]
--[=[
    @prop Position nlerpable
    
    The current position (rotation) at the given clock time. 
    Assigning the position will change the spring to have that position.
--]=]
--[=[
    @prop p nlerpable
    @alias Position
--]=]
--[=[
    @prop Velocity nlerpable
    
    The current velocity. Assigning the velocity will change the spring to have 
    that velocity. The velocity should be in the axis with angle magnitude
    format, where the magnitude represents the angle, and the unit of that
    vector (if magnitude is > 0) represents the axis of rotation.
    The zero vector represents no velocity.
--]=]
--[=[
    @prop v nlerpable
    @alias Velocity
--]=]
--[=[
    @prop Target nlerpable
    
    The current target. Assigning the target will change the spring to have 
    that target.
--]=]
--[=[
    @prop t nlerpable
    @alias Target
--]=]
--[=[
    @prop Damping number
    
    The current damper, defaults to 1. At 1 the spring is critically damped. 
    At less than 1, it will be underdamped, and thus, bounce, and at over 1, 
    it will be critically damped.
--]=]
--[=[
    @prop d number
    @alias Damping
--]=]
--[=[
    @prop Speed number
    
    The speed, defaults to 1, but should be between [0, infinity)
--]=]
--[=[
    @prop s number
    @alias Speed
--]=]
--[=[
    @prop Clock
]=]


--[=[
    @function
    @group Constructors
    
    Constructs a new Quaternion Spring at the position and target specified.
]=]
function QuaternionSpring.new(initial: Quaternion?, damping: number?, speed: number?, clock: (() -> number)?): QuaternionSpring
	local l_initial = initial or Quaternion.identity
	local l_damping = damping or 1
	local l_speed = speed or 1
	local l_clock = clock or os.clock
	
	local self = {
		_clock = l_clock,
		_time = l_clock(),
		_position = l_initial,
		_velocity = Vector3.zero,
		_target = l_initial,
		_damping = l_damping,
		_speed = l_speed,
		_initial = l_initial
	}
	
	return setmetatable(self :: t_QuaternionSpring, QuaternionSpring)
end

--[=[
    @method
    @group Methods
    
    Resets the springs' position and target to the target value provided, or 
    to the initial value the spring was created with if target is not specified.
    Sets the velocity to zero.
]=]

local function Reset(self: QuaternionSpring, target: Quaternion?)
	local setTo = target or self._initial
	self._position = setTo
	self._target = setTo
	self._velocity = Vector3.zero
end

QuaternionSpring.Reset = Reset

--[=[
    @method
    @group Methods
    
    Impulses the spring, increasing velocity by the amount given.
    This is useful to make something shake. Note that the velocity
    will be a rotation vector, such that the axis is the direction
    and the magnitude is the angle (compact axis-angle).
]=]




local function Impulse(self: QuaternionSpring, velocity: Vector3)
	self._velocity = self._velocity + velocity
end

QuaternionSpring.Impulse = Impulse


local function _positionVelocity(self: QuaternionSpring, now: number)
	local currentRotation = self._position
	local currentVelocity = self._velocity
	local targetRotation = self._target
	local dampingFactor = self._damping
	local speed = self._speed
	
	local deltaTime = speed * (now - self._time)
	local dampingSquared = dampingFactor * dampingFactor
	
	local angFreq, sinTheta, cosTheta
	if dampingSquared < 1 then
		angFreq = math.sqrt(1 - dampingSquared)
		local exponential = math.exp(-dampingFactor * deltaTime) / angFreq
		cosTheta = exponential * math.cos(angFreq * deltaTime)
		sinTheta = exponential * math.sin(angFreq * deltaTime)
	elseif dampingSquared == 1 then
		angFreq = 1
		local exponential = math.exp(-dampingFactor * deltaTime)
		cosTheta, sinTheta = exponential, exponential * deltaTime
	else
		angFreq = math.sqrt(dampingSquared - 1)
		local angFreq2 = 2 * angFreq
		local u = math.exp((-dampingFactor + angFreq) * deltaTime) / angFreq2
		local v = math.exp((-dampingFactor - angFreq) * deltaTime) / angFreq2
		cosTheta, sinTheta = u + v, u - v
	end
	
	local pullToTarget = 1 - (angFreq * cosTheta + dampingFactor * sinTheta)
	local velPosPush = sinTheta / speed
	local velPushRate = speed * sinTheta
	local velocityDecay = angFreq * cosTheta - dampingFactor * sinTheta
	
	local posQuat = currentRotation:Slerp(targetRotation, pullToTarget)
	local newPosition = posQuat:Integrate(currentVelocity, velPosPush)
	
	local difQuat = currentRotation:Difference(targetRotation)
	local axis, angle = difQuat:ToAxisAngle()
	local velPush = (axis * angle) * velPushRate
	local velDecay = currentVelocity * velocityDecay
	
	local newVelocity = velPush + velDecay
	
	return newPosition, newVelocity
end

--[=[
    @method
    @group Methods
    Instantly skips the spring forwards by the given time.
]=]
local function TimeSkip(self: QuaternionSpring, delta: number)
	local now = self._clock()
	local position, velocity = _positionVelocity(self, now+delta)
	self._position = position
	self._velocity = velocity
	self._time = now
end

QuaternionSpring.TimeSkip = TimeSkip

function QuaternionSpring.__index(self: QuaternionSpring, index: string)
	if QuaternionSpring[index] then
		return QuaternionSpring[index]
	elseif index == "Position" or index == "p" then
		local position, _ = _positionVelocity(self, self._clock())
		return position
	elseif index == "Velocity" or index == "v" then
		local _, velocity = _positionVelocity(self, self._clock())
		return velocity
	elseif index == "Target" or index == "t" then
		return self._target
	elseif index == "Damping" or index == "d" then
		return self._damping
	elseif index == "Speed" or index == "s" then
		return self._speed
	elseif index == "Clock" then
		return self._clock
	end
	error(string.format(ERROR_FORMAT, tostring(index)), 2)
end

function QuaternionSpring.__newindex(self: QuaternionSpring, index: string, value: any)
	local now = self._clock()
	if index == "Position" or index == "p" then
		local _, velocity = _positionVelocity(self, now)
		self._position = value
		self._velocity = velocity
		self._time = now
	elseif index == "Velocity" or index == "v" then
		local position, _ = _positionVelocity(self, now)
		self._position = position
		self._velocity = value
		self._time = now
	elseif index == "Target" or index == "t" then
		local position, velocity = _positionVelocity(self, now)
		self._position = position
		self._velocity = velocity
		self._target = value
		self._time = now
	elseif index == "Damping" or index == "d" then
		local position, velocity = _positionVelocity(self, now)
		self._position = position
		self._velocity = velocity
		self._damping = value
		self._time = now
	elseif index == "Speed" or index == "s" then
		local position, velocity = _positionVelocity(self, now)
		self._position = position
		self._velocity = velocity
		self._speed = value < 0 and 0 or value
		self._time = now
	elseif index == "Clock" then
		local position, velocity = _positionVelocity(self, now)
		self._position = position
		self._velocity = velocity
		self._clock = value
		self._time = value()
	else
		error(string.format(ERROR_FORMAT, tostring(index)), 2)
	end
end


return QuaternionSpring

File: c/f32/quaternion_spring.c
#include "quaternion_spring.h"

#include <math.h>

#include "quaternion.h"
#include "vector3.h"


// concern: floats may not provide enough accuracy for expf
void evaluate_spring(
    const QuaternionSpring* self, 
    const double now, 
    Quaternion* out_position,
    Vector3* out_velocity
) {
    const Quaternion current_position = self->position;
    const Quaternion current_target = self->target;
    const Vector3 current_velocity = self->velocity;
    const float damping = self->damping;
    const float speed = self->speed;
    
    const float dt = speed * (float)(now - self->_time);
    const float damping_squared = damping * damping;
    
    float ang_freq, sin_theta, cos_theta;
    if (damping_squared < 1) {
        ang_freq = sqrtf(1 - damping_squared);
        const float exponential = expf(-damping * dt) / ang_freq;
        const float afdt = ang_freq * dt;
        #ifdef _GNU_SOURCE
            float sin_tm, cos_tm;
            sincosf(afdt, &sin_tm, &cos_tm);
            sin_theta = exponential * sin_tm;
            cos_theta = exponential * cos_tm;
        #else
            sin_theta = exponential * sinf(afdt);
            cos_theta = exponential * cosf(afdt);
        #endif
    } else if (damping_squared == 1) {
        ang_freq = 1;
        const float exponential = expf(-damping * dt);
        sin_theta = exponential * dt;
        cos_theta = exponential;
    } else {
        ang_freq = sqrtf(damping_squared - 1);
        const float ang_freq_2 = 1 / (2 * ang_freq);
        const float m_damping = -damping;
        const float u = expf((m_damping + ang_freq) * dt) * ang_freq_2;
        const float v = expf((m_damping - ang_freq) * dt) * ang_freq_2;
        sin_theta = u - v;
        cos_theta = u + v;
    }
    
    if (out_position) {
        const float pull_to_target = 1 - (
            ang_freq * cos_theta + damping * sin_theta
        );
        const float vel_pos_push = sin_theta / speed;
        
        const Quaternion pos_quat = quaternion_slerp(
            &current_position, &current_target, pull_to_target
        );
        *out_position = quaternion_integrate(
            &pos_quat, &current_velocity, vel_pos_push
        );
    }
    
    if (out_velocity) {
        const float vel_push_rate = speed * sin_theta;
        const float velocity_decay = ang_freq * cos_theta - damping * sin_theta;
        const Quaternion dif_quat = quaternion_difference(
            &current_position, &current_target
        );
        const Vector3 euler_vec = quaternion_to_euler_vector(&dif_quat);
        const Vector3 vel_push = vector3_scale(&euler_vec, vel_push_rate);
        const Vector3 vel_decay = vector3_scale(&current_velocity, velocity_decay);
    
        *out_velocity = vector3_add(&vel_push, &vel_decay);
    }
}

inline double time_now(const QuaternionSpring* self) {
    return self->clock(self->clock_state);
}


void quaternion_spring_evaluate(
    QuaternionSpring* self,
    Quaternion* out_position,
    Vector3* out_velocity
) {
    double now = time_now(self);
    evaluate_spring(self, now, out_position, out_velocity);
    self->_time = now;
}

void quaternion_spring_evaluate_npv(
    QuaternionSpring* self
) {
    double now = time_now(self);
    self->_time = now;
}


void quaternion_spring_set_position(
    QuaternionSpring* self,
    const Quaternion* position
) {
    double now = time_now(self);
    evaluate_spring(self, now, NULL, &(self->velocity));
    self->position = *position;
    self->_time = now;
}


void quaternion_spring_set_target(
    QuaternionSpring* self,
    const Quaternion* target
) {
    double now = time_now(self);
    evaluate_spring(self, now, &(self->position), &(self->velocity));
    self->target = *target;
    self->_time = now;
}


void quaternion_spring_set_velocity(
    QuaternionSpring* self,
    const Vector3* velocity
) {
    double now = time_now(self);
    evaluate_spring(self, now, &(self->position), NULL);
    self->velocity = *velocity;
    self->_time = now;
}


void quaternion_spring_set_damping(
    QuaternionSpring* self,
    const float damping
) {
    double now = time_now(self);
    evaluate_spring(self, now, &(self->position), &(self->velocity));
    self->damping = damping;
    self->_time = now;
}


void quaternion_spring_set_speed(
    QuaternionSpring* self,
    const float speed
) {
    double now = time_now(self);
    evaluate_spring(self, now, &(self->position), &(self->velocity));
    self->speed = speed;
    self->_time = now;
}

void quaternion_spring_set_clock(
    QuaternionSpring* self,
    const double (*clock)(void*),
    const void* clock_state
) {
    double now = time_now(self);
    evaluate_spring(self, now, &(self->position), &(self->velocity));
    self->clock = clock;
    self->_time = clock(clock_state);
}


void quaternion_spring_reset(
    QuaternionSpring* self,
    const Quaternion* optional_target
) {
    Quaternion target = optional_target ? *optional_target : self->_initial;
    self->_initial = target;
    self->position = target;
    self->target = target;
    self->velocity = VECTOR3_ZERO;
}


void quaternion_spring_impulse(
    QuaternionSpring* self,
    const Vector3* impulse
) {
    self->velocity = vector3_add(&(self->velocity), impulse);
}


void quaternion_spring_time_skip(
    QuaternionSpring* self,
    const double delta
) {
    double now = time_now(self);
    evaluate_spring(self, now + delta, &(self->position), &(self->velocity));
    self->_time = now;
}

